src/main/java/bio/terra/service/snapshot/SnapshotService.java

```diff
@@ -172,8 +172,7 @@ public SnapshotService(
    */
   public String createSnapshot(
       SnapshotRequestModel snapshotRequestModel, AuthenticatedUserRequest userReq) {
-    String sourceDatasetName = snapshotRequestModel.getContents().get(0).getDatasetName();
-    Dataset dataset = datasetService.retrieveByName(sourceDatasetName);
+    Dataset dataset = getSourceDatasetsFromSnapshotRequest(snapshotRequestModel).get(0);
```

Comment:
Now this is done both here and in `SnapshotCreateFlight`. If we need it in both places (and this is the only place `SnapshotCreateFlight` is called), then maybe `Dataset` should be a flight parameter.

Also it looks like this work is repeated in the code that calls this, in `SnapshotApiController.createSnapshot()`:

```
    List<UUID> snapshotSourceDatasetIds =
        snapshotService.getSourceDatasetIdsFromSnapshotRequest(snapshotRequestModel);
```

Reply:
Honestly, it looks like there's even a todo to move the auth check to the flight? The `Dataset` here is only used to set the default billing profile if there isn't one. I think this code could use some serious refactoring. Happy to get into it, but I think that might actually be pretty high scope creep because I would want to push a lot of this in there. Maybe tomorrow we can figure out what changes actually make sense to do? Or what work we are going to push if we want to prioritize the refactor. (Which might mean reprioritizing v2 createSnapshot)

Reply:
I poked around with v2, definitely I would say is out of scope - that's a pretty serious refactor of how we do things. I am going to try to push some of the steps that are done in the controller and service into the job though, which might help.

Reply:
Would you take a look at https://github.com/DataBiosphere/jade-data-repo/pull/1705 when you have the chance? I think that's the real fix here.

Reply:
Discussed offline - I think we have more design to happen here.